Build and add applets in Section 5.9

douden opened the pull request:
@bewuwy & @zlumi

In [5.9. Approximating integrals](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable%20functions/Integration/approximating_integrals.html) you can find a table with figures that should become applets.

Here is the list of relevant figures:

- [x] [Fig. 5.9.2](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable_functions/Integration/approximating_integrals.html#fig-approximatingintegrals-riemann) - https://openla.ewi.tudelft.nl/applet/calculus/riemann_sum/riemann_sum_2d
- [x] [Fig. 5.9.3](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable%20functions/Integration/approximating_integrals.html#fig-numericalintegration-leftrectangle) - https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods
- [x] [Fig. 5.9.4](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable%20functions/Integration/approximating_integrals.html#fig-numericalintegration-trapezoidal) - https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods?rule=trapezoid
- [x] [Fig. 5.9.5](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable%20functions/Integration/approximating_integrals.html#fig-numericalintegration-simpsons) - https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods?rule=simpson
- [x] [Fig 5.9.6] Custom applet for integration rules: https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?xR=2%5Cpi&xL=%5Cfrac%7B%5Cpi%7D%7B4%7D&function=%5Csin%20(x)
- [x] [Fig. 5.9.7](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable_functions/Integration/approximating_integrals.html#fig-approximatingintegrals-composite) - https://openla.ewi.tudelft.nl/applet/calculus/riemann_sum/riemann_sum_2d
- [x] [Fig. 5.9.8](https://tudelft-prime-books.github.io/Calculus/applets_section_5.9/Single-variable%20functions/Integration/approximating_integrals.html#fig-numericalintegration-errors) -https://openla.ewi.tudelft.nl/applet/calculus/integration/method_errors

3D riemann sum: https://openla.ewi.tudelft.nl/applet/calculus/riemann_sum/riemann_sum_3d

If you need more information, please tell me here.

bewuwy commented:
Should we also create a MR in this repository to add the finished applets to the book or do you want to do it?

douden commented:
If an applet is finished I will add it in the branch of this PR to the book 😉

bewuwy commented:
We will probably move the applet to a different path, since we want to split the LA applets in a better way from Calc applets, but I will send you the new URL when that happens.

bewuwy commented:
> We will probably move the applet to a different path, since we want to split the LA applets in a better way from Calc applets, but I will send you the new URL when that happens.

The new URL is https://openla.ewi.tudelft.nl/applet/calculus/integration/left_rectangle

bewuwy commented:
Here are the points the lecturers made at the PGG meeting today about the left rectangle rule applet:

- It would be more interesting to have a not monotone function, like x^3
- Maybe a dropdown for choosing a function
- Should the function have an asymptote?

What do you think @douden ?

douden commented:
1. Monotone is better from a didactic view. Also the function is irrelevant, so no.
2. No.
3. As long as you cannot choose the asymptote, it does not matter.

Only thing that could be added is a dropdown to choose the rule shown, I.e. also right Rectangle and Trapezoidal.

douden commented:
[like]  Dennis den Ouden-van der Horst reacted to your message:
________________________________
From: Bartek ***@***.***>
Sent: Thursday, October 30, 2025 1:29:02 PM
To: TUDelft-PRIME-Books/Calculus ***@***.***>
Cc: Dennis den Ouden-van der Horst ***@***.***>; Author ***@***.***>
Subject: Re: [TUDelft-PRIME-Books/Calculus] Build and add applets in Section 5.9 (PR #14)

[https://avatars.githubusercontent.com/u/61749397?s=20&v=4]bewuwy left a comment (TUDelft-PRIME-Books/Calculus#14)<https://github.com/TUDelft-PRIME-Books/Calculus/pull/14#issuecomment-3468003296>

We will probably move the applet to a different path, since we want to split the LA applets in a better way from Calc applets, but I will send you the new URL when that happens.

The new URL is https://openla.ewi.tudelft.nl/applet/calculus/integration/left_rectangle

—
Reply to this email directly, view it on GitHub<https://github.com/TUDelft-PRIME-Books/Calculus/pull/14#issuecomment-3468003296>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALYF4EW5NXYKYE5QWOJ6CLD32IHB5AVCNFSM6AAAAACKAGDDBGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTINRYGAYDGMRZGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

bewuwy commented:
I created a single applet with all 3 of the rules as a dropdown

https://deploy-preview-331--openla-preview.netlify.app/applet/calculus/integration/simple_methods

The default rule can be specified with a query parameter `rule` - set to either 'trapezoid' or 'simpson' (otherwise rectangle is the default)

So I also think it might be better to just replace the left rectangle URL for this one?

douden commented:
> I created a single applet with all 3 of the rules as a dropdown
> 
> https://deploy-preview-331--openla-preview.netlify.app/applet/calculus/integration/simple_methods
> 
> The default rule can be specified with a query parameter `rule` - set to either 'trapezoid' or 'simpson' (otherwise rectangle is the default)
> 
> So I also think it might be better to just replace the left rectangle URL for this one?

Smart!

Is it possible to:

- [x] rename 'rectangle' to 'left'?
- [x] also include the right Rectangle rule, with name 'right'?
- [x] also include the Midpoint rule, with name 'midpoint'?

Then we would have an all-in-one applet 😄 

The default can still be the left Rectangle rule.

bewuwy commented:
Done, https://deploy-preview-331--openla-preview.netlify.app/applet/calculus/integration/simple_methods

and `rule` param can be either: left, right, trapezoid, midpoint, simpson

douden commented:
Looks and works as it should. If it was up to me: merge it 😉  (the new applet, not this PR)

douden committed:
Merge pull request #53 from TUDelft-PRIME-Books/development

Development

douden committed:
Update figure references and embed applets for integration rules

Replaces outdated figure references with new ones for integration methods and updates static images to interactive applets for the Trapezoidal and Simpson's rules. Also adjusts applet URLs and improves the interactivity table to reflect current implementation status.

bewuwy commented:
Hi, I created the applet for 5.9.7 - https://deploy-preview-338--openla-preview.netlify.app/applet/calculus/integration/method_errors
What do you think @douden ?

douden commented:
Looking good!

Some things that I would like to see improved:
- [x] can you make sure the integral bounds are shown in the integral?
- [x] can you make sure (if not already) that the width of the interval of integration divided by h is always an integer?
- [x] can you make the stationary dots transparent if the animation runs, and the animated dots solid in that case?
- [x] can you add toggles to show/hide the different options?

bewuwy commented:
All done: https://deploy-preview-338--openla-preview.netlify.app/applet/calculus/integration/method_errors

douden commented:
Nice, looking good!

The last thing I noticed is that, on my screen, the _h_ above the horizontal axis is not shown entirely, as it is cut-off by the viewport. Is it possible to (automatically) set the viewport so that the axes labels are at least always shown?

bewuwy commented:
> The last thing I noticed is that, on my screen, the _h_ above the horizontal axis is not shown entirely, as it is cut-off by the viewport. Is it possible to (automatically) set the viewport so that the axes labels are at least always shown?

I was thinking about this issue and I think it's not that easy - we could either go in the geogebra/desmos direction in which the axis labels are sticking to the screen border or keep it as is and just zoom it out slightly in this case.

Example from desmos:
<img width="761" height="414" alt="image" src="https://github.com/user-attachments/assets/7ce162aa-999f-4b09-89cd-39a9ee1c17a4" />

Which option do you prefer? For now I just changed the initial zoom.

bewuwy commented:
I have also created a control for inputting a custom function: https://deploy-preview-339--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom

douden commented:
> > The last thing I noticed is that, on my screen, the _h_ above the horizontal axis is not shown entirely, as it is cut-off by the viewport. Is it possible to (automatically) set the viewport so that the axes labels are at least always shown?
> 
> I was thinking about this issue and I think it's not that easy - we could either go in the geogebra/desmos direction in which the axis labels are sticking to the screen border or keep it as is and just zoom it out slightly in this case.
> 
> Example from desmos:
> <img width="761" height="414" alt="image" src="https://github.com/user-attachments/assets/7ce162aa-999f-4b09-89cd-39a9ee1c17a4" />
> 
> Which option do you prefer? For now I just changed the initial zoom.

I would say that a solution that works for any screen of any size would be best. So something like the desmos solution would be my preference.

douden commented:
> I have also created a control for inputting a custom function: https://deploy-preview-339--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom

That looks good.

Is it possible for me to set the starting view by adding parameters to the url?

I.e, can I set
- [x] the function
- [x] the rule
- [x] the interval

bewuwy commented:
Done, you need to have the `function` string be a latex string. For example:

https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=\sin{(x)}&rule=simpson&xL=0&xR=3.14

You can set bounds with `xR` and `xL`.

bewuwy commented:
Also, Happy Holidays! and New Years!

douden commented:
> Also, Happy Holidays! and New Years!

Same to you!!!!

douden commented:
> Done, you need to have the `function` string be a latex string. For example:
> 
> https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=\sin{(x)}&rule=simpson&xL=0&xR=3.14
> 
> You can set bounds with `xR` and `xL`.

Sounds good. I will test next week.

Can you also merge the error applet and set the url in the initial PR comment to the correct one?

douden commented:
> > Done, you need to have the `function` string be a latex string. For example:
> > [https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=\sin{(x)}&rule=simpson&xL=0&xR=3.14](https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=%5Csin%7B(x)%7D&rule=simpson&xL=0&xR=3.14)
> > You can set bounds with `xR` and `xL`.

Just dit a test. Is it possible to also allow xR and xL to be LaTeX and/or Python strings that evaluate to a number? That would allow also to use `\pi/2` and/or `np.pi/2`.

bewuwy commented:
> > > Done, you need to have the `function` string be a latex string. For example:
> > > [https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=\sin{(x)}&rule=simpson&xL=0&xR=3.14](https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?function=%5Csin%7B(x)%7D&rule=simpson&xL=0&xR=3.14)
> > > You can set bounds with `xR` and `xL`.
> 
> Just dit a test. Is it possible to also allow xR and xL to be LaTeX and/or Python strings that evaluate to a number? That would allow also to use `\pi/2` and/or `np.pi/2`.

Done, here is an example:
https://deploy-preview-351--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom?xR=2\pi&xL=\frac{\pi}{4}&function=\sin%20(x)

The parameters have to be latex strings.

douden commented:
Looks and works good.

Could you

- [x] Merge the error one: https://deploy-preview-338--openla-preview.netlify.app/applet/calculus/integration/method_errors
- [x] Merge the new custom one: https://deploy-preview-351--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom
- [x] In the top comment of this PR add the relevant links?

bewuwy commented:
Yes, I always merge them asap, but sometimes I need to wait for approves ;)

bewuwy commented:
In the merged error one (https://openla.ewi.tudelft.nl/applet/calculus/integration/method_errors) I added the new axis labels that move with the view. I still need to make sure they work better with zooming....

douden commented:
> In the merged error one (https://openla.ewi.tudelft.nl/applet/calculus/integration/method_errors) I added the new axis labels that move with the view. I still need to make sure they work better with zooming....

Smart solution. If the zooming is fixed, even better.

Othweise the "glue" to the intersection of an axis and the screen border would be also an option.

douden commented:
For the error applet I now see this:
<img width="710" height="503" alt="image" src="https://github.com/user-attachments/assets/6c7cf601-b519-4d2c-b103-502775f87a8b" />

I think the original way and design of doing the axis labels is better. Could you revert it back (until a better way is implemented)?

douden commented:
For the simple_methods applet: could you add the horzontal axis label $x$ and the vertical axis label $f(x)$ ?

For the simple_methods_custom applet: could you add the default axes labels as above, but also provide a way to add them to the url and change them in the applet?
So basically for this custom applet I want to change the part $f(x)$ to for example $g(y)$ and I want then to be able to use $y$ in the function description and also see the y and the g in the applet appear at the correct places.

Another point of interest: some characters are not treated correctly when in the url, such as +. MWE that fails: https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods_custom?xR=\frac{\pi}{4}&xL=0&function=\sqrt{1+\cos(x)^2}

Thye + disappears. Using %2B instead of + in the url works.

Could you also look into setting the view of the applet automatically, such that [xl,xr] is in the middle, and the entire view is 2.5 times the length of [xl,xr] ?

And add more decimals to the values of the integral and the approximation? I prefer at least 7.

douden committed:
Refine integration notation and add applets for interactivity

Standardized integral notation throughout the approximating integrals chapter, replacing 't' with 'x' where appropriate and consistently using '\,dx'. Added and updated interactive applets for visualizing integration rules and errors. Updated Grasple exercise links and labels for clarity. Added a new reference for the TeachBooks logo in references.bib.

douden committed:
Update requirements.txt

bewuwy commented:
I agree that the new labels are still not very usable, especially on smaller screens. I will work on them to be more responsive and for now revert to the old ones.

I set the integral and approximation values to be rounded to 7 decimal points.

I think that unfortunately the quirk with '+' sign is unfixable, since it's reserved in the URL for some syntax.

For changing the letters used for the horizontal axis and function you can use URL parameters: `xAxisLetter` and `functionLetter`. Example in the deploy preview: https://deploy-preview-355--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom?xAxisLetter=y&functionLetter=g

bewuwy commented:
The applet view scales now with the boundaries. The axis labels dont make much sense now, but after I work on improving the dynamic ones, it will work better.

Since we have so many URL params, I have also added a table with all the URL parameters when you go into the Share dialog -> embed interactively, for documentation purposes.

douden commented:
> I agree that the new labels are still not very usable, especially on smaller screens. I will work on them to be more responsive and for now revert to the old ones.

Sounds like a good plan and a nice challenge 😉 
 
> I set the integral and approximation values to be rounded to 7 decimal points.

Thanks!

> I think that unfortunately the quirk with '+' sign is unfixable, since it's reserved in the URL for some syntax.

I thought so too. I will open an issue at [Sphinx PRIME applets](https://github.com/TeachBooks/Sphinx-PRIME-applets) so we can convert any special characters in the url so the user can at least use it.
 
> For changing the letters used for the horizontal axis and function you can use URL parameters: `xAxisLetter` and `functionLetter`. Example in the deploy preview: https://deploy-preview-355--openla-preview.netlify.app/applet/calculus/integration/simple_methods_custom?xAxisLetter=y&functionLetter=g

Perfect!

> The applet view scales now with the boundaries. The axis labels dont make much sense now, but after I work on improving the dynamic ones, it will work better.

Let me know!

> Since we have so many URL params, I have also added a table with all the URL parameters when you go into the Share dialog -> embed interactively, for documentation purposes.

That is very handy. Is it possible to also generate the link including all current values set/adjusted by the user?

bewuwy commented:
Thanks for the feedback!

> Is it possible to also generate the link including all current values set/adjusted by the user?

I will try adding that

bewuwy commented:
https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d

That's the riemann sum applet that Zhenghan worked on. I don't know whether it works better for Fig 2 or 7? Should I create a slight variation of it with one closer to 2 and one closer to 7?

 He also created a 3d variant: https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_3d

bewuwy commented:
> Is it possible to also generate the link including all current values set/adjusted by the user?

I added this option. Under the table explaining each parameter there is a text area with the URL with all of the custom params set to current values. Unfortunately it doesnt work for the function input one, because of the way that it stores the latex formula and doesn't expose it..

douden commented:
> https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d
> 
> That's the riemann sum applet that Zhenghan worked on. I don't know whether it works better for Fig 2 or 7? Should I create a slight variation of it with one closer to 2 and one closer to 7?
> 
> He also created a 3d variant: https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_3d

For the 2D one, the concept is fine, but I would like to have
- [x] it more like the https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods in style, including the legend. So that it does not show the explicit function, only the bounds and the calculated area and exact integral value
- [x] The red slider changes the value of dx, and the blue slider changes the value n. That is weird, as they are linked through the formula dx=(b-a)/n. Can you remove the slider for dx?
- [x] I would like to see the same function as in https://openla.ewi.tudelft.nl/applet/calculus/integration/simple_methods.
- [x] Please add draggable points at points a and b, so that we can drag those.

In this way, we can reuse the apploet for both figures.

If you could also make a general version so a user can input/change stuff, that would be awesome!

If you could do something similar for the 3D one (with extremely low priority), we could perhaps use it within another chapter.

douden commented:
> > Is it possible to also generate the link including all current values set/adjusted by the user?
> 
> I added this option. Under the table explaining each parameter there is a text area with the URL with all of the custom params set to current values. Unfortunately it doesnt work for the function input one, because of the way that it stores the latex formula and doesn't expose it..

Would it be an option, for the function input, to have a shadow variable that has the encoded latex string?

bewuwy commented:
You're right. I did it like this and now it works :)

bewuwy commented:
Hi, I made the changes you requested to the riemann sum applet. I was not sure whether you wanted the style to be exactly the same as the simple_methods one, so I created two variants for now.

https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d

https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d_alt


Which one do you prefer? Also, should I change the T_i to x*_i ?

I have also created a custom riemann sum applet: https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d_custom?function=y^2-y^3&n=8&xL=-1.50&xR=4.50&xAxisLetter=y&functionLetter=g

bewuwy commented:
I have also improved the dynamic axis labels. What do you think?

https://deploy-preview-370--openla-preview.netlify.app/applet/calculus/integration/method_errors

https://deploy-preview-370--openla-preview.netlify.app/applet/calculus/integration/simple_methods

douden commented:
I like this style (I prefer consistency)

> https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d_alt

> Which one do you prefer? Also, should I change the T_i to x*_i ?

Yes, definitely change that.

> I have also created a custom riemann sum applet: [https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d_custom?function=y^2-y^3&n=8&xL=-1.50&xR=4.50&xAxisLetter=y&functionLetter=g](https://deploy-preview-310--openla-preview.netlify.app/applet/calculus/riemann_sum/riemann_sum_2d_custom?function=y%5E2-y%5E3&n=8&xL=-1.50&xR=4.50&xAxisLetter=y&functionLetter=g)

That looks so cool. And I think that the idea of creating general versions would save a huge amount of time for all of us.

> I have also improved the dynamic axis labels. What do you think?

> https://deploy-preview-370--openla-preview.netlify.app/applet/calculus/integration/method_errors

When zooming in, the labels disappear.

> https://deploy-preview-370--openla-preview.netlify.app/applet/calculus/integration/simple_methods

I like this one the most, as it (as far as I tested) keeps every label visible.

bewuwy commented:
Thanks, I changed the riemann sum applet to the consistent variant. I also changed the points to be x*. So I think those applets are done?

I'm glad you like the new axis labels. It's true that in the method errors the labels dissapear quickly when zooming in. I will try to fix it. It's trying to keep the labels in the center of the axis.

douden commented:
> Thanks, I changed the riemann sum applet to the consistent variant. I also changed the points to be x*. So I think those applets are done?

I do think so too. Can you finalise and merge the applets?
 
> I'm glad you like the new axis labels. It's true that in the method errors the labels dissapear quickly when zooming in. I will try to fix it. It's trying to keep the labels in the center of the axis.

Sounds like a plan!

douden committed:
Convert integration figures to interactive applets

Remove the admonition listing figures-to-implement and replace several static SVG figures with interactive applet blocks (Riemann, composite illustrations and a composite spoiler applet). Rename the Riemann intermediate points reference from T_n to x_k^*. Add explanatory text for the new spoiler applet, adjust figure attributes (remove figwidth where applicable, keep dark-light class), change British spelling "summarise", and update the Grasple embed ID and label to the new exercise. These changes enable interactive visualizations and update related text and embeds.

douden committed:
Merge branch 'development' into applets_section_5.9
